File: Binary Tree Paths.cpp
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param root the root of the binary tree
     * @return all root-to-leaf paths
     */
    vector<string> binaryTreePaths(TreeNode* root) {
        // Write your code here
        string path;
        vector<string> ret;
        if (!root)
        {
            return ret;
        }
        doHelper(root, path, ret);
        return ret;
    }
    
    void doHelper(TreeNode *root, string s, vector<string> &ret)
    {
        if (!root->right && !root->left)
        {
            ret.emplace_back(s + to_string(root->val));
            return;
        }
        else
        {
            if (root->right)
            {
                doHelper(root->right, s + to_string(root->val) + "->", ret);    
            }
            if (root->left)
            {
                doHelper(root->left, s + to_string(root->val) + "->", ret);    
            }
            
        }
    }
};

File: Binary Tree Paths.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""
class Solution:
    # @param {TreeNode} root the root of the binary tree
    # @return {List[str]} all root-to-leaf paths
    def binaryTreePaths(self, root):
        # Write your code here
        
        
        def helper(node, tmp, ret):
            # if not node:
            #     return
            # else:
            #     tmp.append(str(node.val))
            #     if not node.left and not node.right:
            #         ret.append('->'.join(tmp))
            #     helper(node.left, tmp, ret)
            #     helper(node.right, tmp, ret)
            #     tmp.pop()
            if not node:
                return
            if not node.left and not node.right:
                ret.append(tmp)
            if node.left:
                helper(node.left, tmp + '->' + str(node.left.val), ret)
            if node.right:
                helper(node.right, tmp + '->' + str(node.right.val), ret)
                
        if not root:
            return []
        tmp, ret = str(root.val), []
        helper(root, tmp, ret)
        return ret
